src/librustc/ty/maps/plumbing.rs

```diff
@@ -492,3 +602,253 @@ macro_rules! define_provider_struct {
         }
     };
 }
+
+pub fn force_from_dep_node<'a, 'gcx, 'lcx>(tcx: TyCtxt<'a, 'gcx, 'lcx>,
+                                           dep_node: &DepNode)
+                                           -> bool {
+    use ty::maps::keys::Key;
+    use hir::def_id::LOCAL_CRATE;
+
+    // We should never get into the situation of having to force this from the DepNode.
+    // Since we cannot reconstruct the query key, we would always end up having to evaluate
+    // the first caller of this query that *is* reconstructible. This might very well be
+    // compile_codegen_unit() in which case we'd lose all re-use.
+    debug_assert!(dep_node.kind != DepKind::CodegenUnit,
+                  "calling force_from_dep_node() on DepKind::CodegenUnit");
+
+    if !dep_node.kind.can_reconstruct_query_key() {
+        return false
+    }
+
+    macro_rules! def_id {
+        () => {
+            if let Some(def_id) = dep_node.extract_def_id(tcx) {
+                def_id
+            } else {
+                // return from the whole function
+                return false
+            }
+        }
+    };
+
+    macro_rules! krate {
+        () => { (def_id!()).krate }
+    };
+
+    macro_rules! force {
+        ($query:ident, $key:expr) => {
+            {
+                use $crate::util::common::{ProfileQueriesMsg, profq_msg};
+
+                // FIXME(eddyb) Get more valid Span's on queries.
+                // def_span guard is necessary to prevent a recursive loop,
+                // default_span calls def_span query internally.
+                let span = if stringify!($query) != "def_span" {
+                    $key.default_span(tcx)
+                } else {
+                    ::syntax_pos::DUMMY_SP
+                };
+
+                profq_msg!(tcx,
+                    ProfileQueriesMsg::QueryBegin(
+                        span.clone(),
```

Comment:
Should be `span.data()`. Why needs the `clone()`?

```
[00:06:30] error[E0308]: mismatched types
[00:06:30]    --> /checkout/src/librustc/ty/maps/plumbing.rs:654:25
[00:06:30]     |
[00:06:30] 654 |                         span.clone(),
[00:06:30]     |                         ^^^^^^^^^^^^ expected struct `syntax_pos::SpanData`, found struct `syntax_pos::Span`
[00:06:30] ...
[00:06:30] 710 |         DepKind::RegionScopeTree => { force!(region_scope_tree, def_id!()); }
[00:06:30]     |                                       ------------------------------------- in this macro invocation
[00:06:30]     |
[00:06:30]     = note: expected type `syntax_pos::SpanData`
[00:06:30]                found type `syntax_pos::Span`
[00:06:30]     = help: here are some functions which might fulfill your needs:
[00:06:30]             - .data()
...
[00:06:32] error: aborting due to 110 previous errors
[00:06:32] 
[00:06:32] error: Could not compile `rustc`.
```

Reply:
Looks like this needs a rebase...